docs/design/02-topologies.md

```diff
@@ -0,0 +1,288 @@
+
+- Repo configs will be able to use a new object in the magic `alces` namespace,
+  `alces.config.$question_identifier`, where `$question_identifier` is the
+  identifier for a particular question specified in `configure.yaml`; when
+  templating the corresponding question answer will then be used in these
+  places.
+
+
+# `configure.yaml` format
+
+There will also be a `configure.yaml` in the repo root, which specifies
+configuration questions for the different categories, with the following
+suggested format:
+
+```yaml
+questions:
+  example_question_identifier: &example_question_identifier
+    question: 'What value should this take?' # Required.
+    choices: # Optional; limit choices to one of these options.
+      - 7
+      - 11
+    default: 11 # Optional; default value to use when no input entered, if any.
+    type: 'integer' # Optional/default `'string'`. For validation and converting; see below for details.
+    required: true # Optional/default `true`
+
+  another_example: &another_example
+    question: 'Enter anything?'
+
+all:
+  # Both these questions will be asked for `metal configure all`.
+  example_question_identifier: *example_question_identifier
+  another_example: *another_example
+
+group:
+  # Just ask this question for `metal configure group`.
+  another_example: *another_example
+
+node:
+  # Similarly, just ask this for `metal configure node`. Note: the answer will
+  # be saved under `different_name` in the resulting config this time however.
+  different_name: *example_question_identifier
+```
+
+## Notes regarding above format
+
+- By defining all the questions within a `questions` object we can then share
+  questions between different categories without duplication using YAML's
+  ability to refer to other parts of the document.
+
+- Any name key can be used when specifying a question for a particular
+  category. Often the same identifier will be used for the question definition
+  and when specifying a question for a category, e.g. a `cluster_name` question
+  may be defined in `questions` and then used as `cluster_name` under `all`,
+  however this will not necessarily always need to be the case.
+
+## Notes on `type`
+
+The `type` for a question will serve the dual purpose of validating input and
+then converting this before saving.
+
+For the minimal version of this feature possible type values will probably just
+include `string`, `integer`, and `boolean`. These will be necessary as since we
+are accepting user input we won't know what type we should save a given answer
+as, and we will need non-string types when templating.
+
+Later we can add types with more validation here, e.g. we can have an `ip` type
+which will save the value as a string but validate that it is a valid IP first.
+While these more complicated validations could be specified separately by
+adding an optional regular expression field to question configs, I think
+building these in to Metalware would be better since:
```

Comment:
Once we can initialize a validator object with the `yaml` file. Because it is likely the user will make a mistake entering an input. The validator will be responsible for confirming the input and prompting the user to reenter it if an error occurs.